### ISSUE TITLE
Failed to compile serialization of std::queue

### ISSUE THREAD

**Issue by fenghou1st** [eyalz800/zpp_bits#91]
zpp_bits version: 4.4.10

The simplified code:
```cpp
class SoaBase {
 public:
  constexpr static auto serialize(auto &archive, SoaBase &self) {
    return archive(self.recycled_indexes_);
  }

  std::queue<uint32_t> recycled_indexes_;
};
```

The error message:
```
C:\Libraries\zpp_bits\zpp_bits.h(1378): error C2672: 'zpp::bits::access::visit_members': no matching overloaded function found
C:\Libraries\zpp_bits\zpp_bits.h(2524): note: see reference to function template instantiation 'decltype(auto) zpp::bits::visit_members<std::queue<uint32_t,std::deque<uint32_t,std::allocator<std::seed_seq::result_type>>>&,zpp::bits::in<_ReturnType,zpp::bits::options::alloc_limit<1073741824>>::serialize_one::<lambda_1>>(_T0,_T1 &&)' being compiled
        with
        [
            _T0=std::queue<uint32_t,std::deque<uint32_t,std::allocator<std::seed_seq::result_type>>> &,
            _T1=zpp::bits::in<_ReturnType,zpp::bits::options::alloc_limit<1073741824>>::serialize_one::<lambda_1>
        ]
...
```

**Comment by eyalz800** [comment]
Try to use std::deque directly

**Comment by fenghou1st** [comment]
Got it. 👌

**Comment by fenghou1st** [comment]
A very weird compilation error. After I changed std::queue to std::deque, I was able to compile successfully. After I changed some serialization methods, I started to get this error. So I restored all serialization method modifications and returned the code to the state after changing to std::deque. But I found that the compilation also failed now, with the same error.

```
C:\Libraries\zpp_bits\zpp_bits.h(2664): error C2760: syntax error: ')' was unexpected here; expected '{'
C:\Libraries\zpp_bits\zpp_bits.h(2509): note: see reference to function template instantiation 'zpp::bits::errc zpp::bits::in<_ReturnType,zpp::bits::options::alloc_limit<1073741824>>::serialize_one<unsigned int,std::deque<uint32_t,std::allocator<std::seed_seq::result_type>>&>(_T0)' being compiled
        with
        [
            _T0=std::deque<uint32_t,std::allocator<std::seed_seq::result_type>> &
        ]
C:\Libraries\zpp_bits\zpp_bits.h(2510): note: see reference to function template instantiation 'zpp::bits::errc zpp::bits::in<_ReturnType,zpp::bits::options::alloc_limit<1073741824>>::serialize_one<unsigned int,std::deque<uint32_t,std::allocator<std::seed_seq::result_type>>&>(_T0)' being compiled
        with
        [
            _T0=std::deque<uint32_t,std::allocator<std::seed_seq::result_type>> &
        ]

...

C:\Libraries\zpp_bits\zpp_bits.h(2664): error C2059: syntax error: ')'
C:\Libraries\zpp_bits\zpp_bits.h(2646): error C2143: syntax error: missing ';' before '{'
C:\Libraries\zpp_bits\zpp_bits.h(2157): error C2760: syntax error: ')' was unexpected here; expected '{'

...

C:\Libraries\zpp_bits\zpp_bits.h(2157): error C2059: syntax error: ')'
C:\Libraries\zpp_bits\zpp_bits.h(2103): error C2143: syntax error: missing ';' before '{'
```

**Comment by eyalz800** [comment]
Check that you didn't accidentally modify the zpp_bits library

**Comment by fenghou1st** [comment]
git didn't find any changes:
```bash
fengh@DESKTOP-MFN6HJJ MINGW64 /c/Libraries/zpp_bits (main)
$ git pull origin main
From https://github.com/eyalz800/zpp_bits
 * branch            main       -> FETCH_HEAD
Already up to date.

fengh@DESKTOP-MFN6HJJ MINGW64 /c/Libraries/zpp_bits (main)
$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean

fengh@DESKTOP-MFN6HJJ MINGW64 /c/Libraries/zpp_bits (main)
$
```

This is a simplified code that you can try locally:

```cpp
#include <deque>
#include <type_traits>
#include <gtest/gtest.h>
#include <zpp_bits.h>

class MyBase {
 public:
  constexpr static auto serialize(auto &archive, auto &self) {
    return archive(self.a_);
  }

  std::deque<uint32_t> a_{};
};

class MyDerived : public MyBase {
 public:
  constexpr static auto serialize(auto &archive, auto &self) {
    return archive(std::common_reference_t<MyBase &, decltype(self)>(self), self.b_);
  }

  int b_{};
};

TEST (DerivedClassSerializationTests, Default) {
  MyDerived ori_derived;
  std::vector<std::byte> data;
  zpp::bits::out out(data);
  out(ori_derived).or_throw();

  //
  MyDerived new_derived;
  zpp::bits::in in(data);
  in(new_derived).or_throw();

  //
  EXPECT_EQ(ori_derived.b_, new_derived.b_);
}
```

**Comment by eyalz800** [comment]
It reproduces here on MSVC but not on clang, I wonder why - https://godbolt.org/z/x4ecM6TYo

**Comment by fenghou1st** [comment]
Well, it's a good thing to be able to identify the problem.

**Comment by eyalz800** [comment]
That's what I'm doing, meanwhile close the other issue it's unrelated

**Comment by fenghou1st** [comment]
OK, I'll check that it compiles without std::deque, and then close the other thread.

**Comment by eyalz800** [comment]
Try this https://github.com/eyalz800/zpp_bits/pull/92

**Comment by eyalz800** [comment]
https://godbolt.org/z/89Y47ed75

**Comment by eyalz800** [comment]
It's definitely a compiler issue but for now I don't have a workaround, this is as close as I got:
https://godbolt.org/z/fh8MMrd83 - line 2105
https://godbolt.org/z/3Mdo1qeeb - same code works, same compiler, same flags

**Comment by fenghou1st** [comment]
I updated to MSVC v19.33, this time getting new compilation errors (whether using #92 or main branch):

```
C:\Libraries\zpp_bits\zpp_bits.h(315): error C7685: there is no type named 'serialize' in 'MyDerived'
D:\cpp\test20\test\derived_class_serialization_tests.cpp(21): note: see declaration of 'MyDerived'
C:\Libraries\zpp_bits\zpp_bits.h(810): note: see reference to function template instantiation 'auto zpp::bits::access::has_protocol<MyDerived&>(void)' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(813): note: see reference to variable template 'bool has_protocol<MyDerived &>' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(2264): note: see reference to variable template 'bool by_protocol<MyDerived &>' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(2366): note: see reference to function template instantiation 'zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_many<MyDerived&,>(_T0)' being compiled
        with
        [
            ByteView=std::vector<std::byte,std::allocator<std::byte>>,
            _T0=MyDerived &
        ]
C:\Libraries\zpp_bits\zpp_bits.h(2370): note: see reference to function template instantiation 'zpp::bits::errc zpp::bits::basic_out<ByteView>::serialize_many<MyDerived&,>(_T0)' being compiled
        with
        [
            ByteView=std::vector<std::byte,std::allocator<std::byte>>,
            _T0=MyDerived &
        ]
D:\cpp\test20\test\derived_class_serialization_tests.cpp(34): note: see reference to function template instantiation 'zpp::bits::errc zpp::bits::out<std::vector<std::byte,std::allocator<std::byte>>>::()<MyDerived&>(MyDerived &)' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(242): error C7685: there is no type named 'serialize' in 'MyDerived'
D:\cpp\test20\test\derived_class_serialization_tests.cpp(21): note: see declaration of 'MyDerived'
C:\Libraries\zpp_bits\zpp_bits.h(745): note: see reference to function template instantiation 'auto zpp::bits::access::has_serialize<MyDerived&,zpp::bits::traits::visitor<MyDerived,std::monostate>>(void)' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(801): note: see reference to variable template 'bool has_serialize<MyDerived &>' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(2237): note: see reference to variable template 'bool bitset<MyDerived &>' being compiled
C:\Libraries\zpp_bits\zpp_bits.h(255): error C7685: there is no type named 'serialize' in 'MyDerived'
D:\cpp\test20\test\derived_class_serialization_tests.cpp(21): note: see declaration of 'MyDerived'
```

It looks like Microsoft's compiler has a lot of problems...

**Comment by fenghou1st** [comment]
To make matters worse, even if I removed std::deque, the compilation still failed.
Looks like I'll have to roll back the MSVC version.

**Comment by fenghou1st** [comment]
MSVC v19.31 with #92 successfully compiled on local machine.

**Comment by eyalz800** [comment]
@fenghou1st I have a better option that has good chance to resolve it - just move this requires clause into a concept, I’ll try it in a few hours

**Comment by eyalz800** [comment]
https://godbolt.org/z/bhh383Y4Y

**Comment by fenghou1st** [comment]
MSVC v19.32 with v4.4.11 successfully compiled on local machine. 👍
MSVC v19.33 still got the same error, but it doesn't matter, I'll stick to v19.32 for a while.

I think this issue can be closed, but it's up to you to decide.

**Comment by eyalz800** [comment]
This issue is fixed however the new issue that you see on 19.33 is unrelated and again looks like a compiler issue
